core/src/main/java/org/aerogear/mobile/core/MobileCore.java

```diff
@@ -45,23 +53,87 @@ public Logger defaultLog() {
         };
     }
 
-    @Override
     public void bootstrap(Object... args) {
 
-
         try (InputStream configStream = context.getAssets().open(this.mobileServiceFileName);) {
+
+        Set<String> servicesInitted = new HashSet<>();
+        Set<String> servicesPending = new HashSet<>();
+
             this.configurationMap = MobileCoreJsonParser.parse(configStream);
+            for (Map.Entry<String, Class<? extends ServiceModule>> serviceModuleEntry : serviceRegistry.services()) {
+                String serviceName = serviceModuleEntry.getKey();
+                servicesPending.add(serviceName);
+            }
+
+            boolean unresolvableServiceDetected = false;//This is a bad name/initialization.  The
+                                                        // goal is to reset this at the beginning of
+                                                        //each iteration of the loop and clear it
+                                                        //when we instanciate a service.  If we have
+                                                        //a loop where we can't start a service
+                                                        //throw an exception.
+            while (!servicesPending.isEmpty()) {
+                unresolvableServiceDetected = true;
+                String serviceUnderInstanciation = "";
+                for (String serviceName : servicesPending) {
+                    List<String> dependencies = serviceRegistry.getDependenciesFor(serviceName);
+                    serviceUnderInstanciation = serviceName;
+                    if (dependencies.isEmpty() || servicesInitted.containsAll(dependencies)) {
+                        unresolvableServiceDetected = false;
+                        Class<? extends ServiceModule> serviceClass = serviceRegistry.getServiceClass(serviceName);
```

Comment:
We could extract a method here called `initService` or `startService` for the sake of clarity. This piece is a bit convoluted.
```java
private void startService(String serviceName) {
    Class<? extends ServiceModule> serviceClass = serviceRegistry.getServiceClass(serviceName);
    ServiceModule serviceInstance = serviceClass.newInstance();
    serviceInstance.bootstrap(this, getConfig(serviceName));
    this.services.put(serviceName, serviceInstance);
    servicesInitted.add(serviceName);
}
```
I also would move sentence in 92 to right after service is started and before it breaks from the loop. The final condition would be a bit clearer:
```java
if (unresolvableServiceDetected) {
    throw new ...
}
```